x/mongo/driver/operation.go

```diff
@@ -1474,7 +1478,49 @@
 	return bsoncore.AppendDocumentElement(dst, "readConcern", data), nil
 }
 
-func (op Operation) addWriteConcern(dst []byte, desc description.SelectedServer) ([]byte, error) {
+func marshalBSONWriteConcern(wc writeconcern.WriteConcern, wtimeout time.Duration) (bson.Type, []byte, error) {
+	var elems []byte
+	if wc.W != nil {
+		// Only support string or int values for W. That aligns with the
+		// documentation and the behavior of other functions, like Acknowledged.
+		switch w := wc.W.(type) {
+		case int:
+			if w < 0 {
+				return 0, nil, errNegativeW
+			}
+
+			// If Journal=true and W=0, return an error because that write
+			// concern is ambiguous.
+			if wc.Journal != nil && *wc.Journal && w == 0 {
+				return 0, nil, errInconsistent
+			}
+
+			elems = bsoncore.AppendInt32Element(elems, "w", int32(w))
```

Comment:
I believe this can be resolved by adding a bounds check on `w`.

E.g.
```go
if w > math.MaxInt32 {
	return 0, nil, fmt.Errorf("write concern `w` value %v overflows int32", w)
}
```